### PR TITLE
Remove display plugin

### PR DESCRIPTION
**Pull request by astrofrog**
This moves the last remaining plugin from astropy core to a standalone plugin (https://github.com/astropy/pytest-astropy-header), and fixes https://github.com/astropy/pytest-astropy/issues/15.

Note that ``pytest_terminal_summary`` is actually a core-specific workaround, so I am moving it to the ``conftest.py`` in core (it's only needed because we do funny stuff with IPython in astropy.utils).

This PR won't be mergeable until a release of pytest-astropy-header and pytest-astropy is made. I'm waiting for feedback before doing that though.

### Benefits to migrating

The main motivation for migrating is that this will make it easier in future to make improvements to the plugin without having to rely on the astropy release cycle. It also should simplify configuration for the plugin in third party packages (packages can easily opt-in/out of this with pytest ini options and/or command-line flags). This was the last plugin that still lived in astropy core.

### Migration for packages

As usual, we have to be careful to think of all testing scenarios when moving a plugin, to minimize pain. I've therefore included a section in the README of the plugin package on migrating to this new plugin. Note that the new plugin is essentially opt-in - packages need to take action to switch to the new one, which is necessary to avoid people seeing the astropy header for unrelated packages. See the [README](https://github.com/astropy/pytest-astropy-header) for more details.

If packages don't do anything, things will continue to work except that a deprecation warning will be raised with astropy 4.0 (and until we decide to break things). If Astropy 4.0 is installed, but pytest-astropy is not recent enough (meaning pytest-astropy-header isn't installed), the header might be missing but nothing should error.

### Impact for core package

There should be no impact for the core package, except that a recent version of pytest-astropy will be required for testing for the header to be present.

### Impact for other packages

For third-party packages, there are several cases to consider. I'm assuming packages follow the migration instructions in the [README](https://github.com/astropy/pytest-astropy-header).

When the plugin is used and astropy<4 is installed, the ``PYTEST_HEADER_MODULES`` and
``TESTED_VERSIONS`` variables are actually shared with astropy. The new plugin here is smart enough to detect if the original plugin is enabled, and if so, the one here will do nothing and the astropy one will work as expected since the variables are shared. For astropy 2.0 the display header was not a real plugin so we can disable it as described using ``del pytest_report_header`` as described in the migration guide.

With astropy >4 this should work as expected since there is now only one plugin, the one in the new plugin package.

Here's an example with ccdproc including the changes described in the migration docs here. so with ``conftest.py`` set to:

```python
import os

from .tests.pytest_fixtures import *

from astropy.version import version as astropy_version
if astropy_version < '3.0':
    from astropy.tests.pytest_plugins import *
    del pytest_report_header
else:
    from pytest_astropy.display import PYTEST_HEADER_MODULES, TESTED_VERSIONS


def pytest_configure(config):

    config.option.astropy_header = True

    PYTEST_HEADER_MODULES.pop('h5py', None)
    PYTEST_HEADER_MODULES['Astropy'] = 'astropy'
    PYTEST_HEADER_MODULES['astroscrappy'] = 'astroscrappy'
    PYTEST_HEADER_MODULES['reproject'] = 'reproject'

    from .version import version, astropy_helpers_version
    packagename = os.path.basename(os.path.dirname(__file__))
    TESTED_VERSIONS[packagename] = version
    TESTED_VERSIONS['astropy_helpers'] = astropy_helpers_version
```

**Running the tests with python setup.py test**

_With astropy 2.0.15:_

```
=================================================================== test session starts ===================================================================
platform linux -- Python 3.7.3, pytest-3.6.4, py-1.8.0, pluggy-0.7.1

Running tests with ccdproc version 2.1.dev1561.
Running tests in ccdproc docs.

Date: 2019-09-09T16:13:55

Platform: Linux-5.0.0-27-generic-x86_64-with-Ubuntu-19.04-disco

Executable: /home/tom/python/dev/bin/python

Full Python Version: 
3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.17.0
Scipy: 1.3.1
Matplotlib: 3.1.1
Pandas: 0.25.0
Astropy: 2.0.14
astroscrappy: 1.0.8
reproject: 0.5
astropy-helpers: 3.2.1
```

_With astropy 3.2.1:_

```
=================================================================== test session starts ===================================================================
platform linux -- Python 3.7.3, pytest-4.6.5, py-1.8.0, pluggy-0.12.0

Running tests with ccdproc version 2.1.dev1561.
Running tests in ccdproc docs.

Date: 2019-09-09T15:57:42

Platform: Linux-5.0.0-27-generic-x86_64-with-Ubuntu-19.04-disco

Executable: /home/tom/python/dev/bin/python

Full Python Version: 
3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Numpy: 1.17.0
Scipy: 1.3.1
Matplotlib: 3.1.1
Pandas: 0.25.0
Astropy: 3.2.1
astroscrappy: 1.0.8
reproject: 0.5
astropy_helpers: 3.2.1
Using Astropy options: remote_data: none.
```

_With astropy dev (after plugin removal, see https://github.com/astropy/astropy/pull/9214):_

```
=================================================================== test session starts ===================================================================
platform linux -- Python 3.7.3, pytest-3.6.4, py-1.8.0, pluggy-0.7.1

Running tests with ccdproc version 2.1.dev1561.
Running tests in ccdproc docs.

Date: 2019-09-09T16:48:17

Platform: Linux-5.0.0-27-generic-x86_64-with-Ubuntu-19.04-disco

Executable: /home/tom/python/dev/bin/python

Full Python Version: 
3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.17.0
Scipy: 1.3.1
Matplotlib: 3.1.1
Pandas: 0.25.0
Astropy: 4.0.dev25644
astroscrappy: 1.0.8
reproject: 0.5
astropy-helpers: 3.2.1

Using Astropy options: remote_data: none.
```

**Running the tests with pytest**

_With astropy 2.0.15:_

```
$ pytest ccdproc
=================================================================== test session starts ===================================================================
platform linux -- Python 3.7.3, pytest-3.6.4, py-1.8.0, pluggy-0.7.1

Running tests with ccdproc version 2.1.dev1561.
Running tests in ccdproc.

Date: 2019-09-09T16:18:58

Platform: Linux-5.0.0-27-generic-x86_64-with-Ubuntu-19.04-disco

Executable: /home/tom/python/dev/bin/python3.7

Full Python Version: 
3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.17.0
Scipy: 1.3.1
Matplotlib: 3.1.1
Pandas: 0.25.0
Astropy: 2.0.14
astroscrappy: 1.0.8
reproject: 0.5
astropy-helpers: 3.2.1

Using Astropy options: remote_data: none.
```

_With astropy 3.2.1:_

```
$ pytest ccdproc
=================================================================== test session starts ===================================================================
platform linux -- Python 3.7.3, pytest-4.6.5, py-1.8.0, pluggy-0.12.0

Running tests with ccdproc version 2.1.dev1561.
Running tests in ccdproc.

Date: 2019-09-09T15:59:06

Platform: Linux-5.0.0-27-generic-x86_64-with-Ubuntu-19.04-disco

Executable: /home/tom/python/dev/bin/python3.7

Full Python Version: 
3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.17.0
Scipy: 1.3.1
Matplotlib: 3.1.1
Pandas: 0.25.0
Astropy: 3.2.1
astroscrappy: 1.0.8
reproject: 0.5
astropy-helpers: 3.2.1
```

_With astropy dev:_

```
$ pytest ccdproc
=================================================================== test session starts ===================================================================
platform linux -- Python 3.7.3, pytest-3.6.4, py-1.8.0, pluggy-0.7.1

Running tests with ccdproc version 2.1.dev1561.
Running tests in ccdproc.

Date: 2019-09-09T16:49:22

Platform: Linux-5.0.0-27-generic-x86_64-with-Ubuntu-19.04-disco

Executable: /home/tom/python/dev/bin/python3.7

Full Python Version: 
3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.17.0
Scipy: 1.3.1
Matplotlib: 3.1.1
Pandas: 0.25.0
Astropy: 4.0.dev25644
astroscrappy: 1.0.8
reproject: 0.5
astropy-helpers: 3.2.1
```

**Running the tests with package.test()**

_With astropy 2.0.14:_

```
In [2]: ccdproc.test()                                                                                                                                     
=================================================================== test session starts ===================================================================
platform linux -- Python 3.7.3, pytest-3.6.4, py-1.8.0, pluggy-0.7.1

Running tests with ccdproc version 2.1.dev1561.
Running tests in ccdproc.

Date: 2019-09-09T16:12:57

Platform: Linux-5.0.0-27-generic-x86_64-with-Ubuntu-19.04-disco

Executable: /home/tom/python/dev/bin/python3.7

Full Python Version: 
3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.17.0
Scipy: 1.3.1
Matplotlib: 3.1.1
Pandas: 0.25.0
Astropy: 2.0.14
astroscrappy: 1.0.8
reproject: 0.5
astropy-helpers: 3.2.1
```

_With astropy 3.2.1:_

```
In [2]: ccdproc.test()                                                                                                                                     
=================================================================== test session starts ===================================================================
platform linux -- Python 3.7.3, pytest-4.6.5, py-1.8.0, pluggy-0.12.0

Running tests with ccdproc version 2.1.dev1561.
Running tests in ccdproc.

Date: 2019-09-09T16:00:41

Platform: Linux-5.0.0-27-generic-x86_64-with-Ubuntu-19.04-disco

Executable: /home/tom/python/dev/bin/python3.7

Full Python Version: 
3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Numpy: 1.17.0
Scipy: 1.3.1
Matplotlib: 3.1.1
Pandas: 0.25.0
Astropy: 3.2.1
astroscrappy: 1.0.8
reproject: 0.5
astropy_helpers: 3.2.1
Using Astropy options: remote_data: none.
```

_With astropy dev:_

```
In [2]: ccdproc.test()                                                                                                                                     
WARNING: AstropyDeprecationWarning: The astropy.tests.plugins.display plugin has been deprecated. See the pytest-astropy documentation for information on migrating to using pytest-astropy to customize the pytest header. [astropy.tests.plugins.display]
=================================================================== test session starts ===================================================================
platform linux -- Python 3.7.3, pytest-3.6.4, py-1.8.0, pluggy-0.7.1

Running tests with ccdproc version 2.1.dev1561.
Running tests in ccdproc.

Date: 2019-09-09T16:49:49

Platform: Linux-5.0.0-27-generic-x86_64-with-Ubuntu-19.04-disco

Executable: /home/tom/python/dev/bin/python3.7

Full Python Version: 
3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.17.0
Scipy: 1.3.1
Matplotlib: 3.1.1
Pandas: 0.25.0
Astropy: 4.0.dev25644
astroscrappy: 1.0.8
reproject: not available
astropy-helpers: 3.2.1

Using Astropy options: remote_data: none.
```

### CODE REVIEW

**Comment by bsipocz** [comment]
💯 for the new plugin

**Comment by pllim** [comment]
Let's do this! 

ohyeahh

**Comment by pllim** [comment]
1. Do a release of the new plugin.
2. Include it in `pytest-astropy`.
3. Re-run the CIs here so it is green.
4. I approve this.
5. This gets merged.
6. Some email out to astropy-dev.
7. Profit!!!

**Comment by bsipocz** [comment]
@astrofrog - I'm happy to help out here to get this in before Friday, let me know.

**Comment by astrofrog** [comment]
I'm going to put aside time this evening to do/finish this but will let you know if I run out of time.

**Comment by astrofrog** [comment]
> We need to release the plugin upstream first, right?

I already released the plugin, so the CI is passing here and this is good to go :) I agree we can do the We need to release the plugin upstream first, right? We can fix that warning in the plugin but that's not feature freeze critical right?

**Comment by bsipocz** [comment]
I agree with @astrofrog, no warnings should hold this up,  and that PR and issue is now upstream rather than FF related :)
Let me have a last look at diff.

File: .circleci/config.yml [hunk 1]
```diff
@@ -71,7 +77,9 @@ jobs:
       - checkout
       - run:
           name: Run tests
-          command: python3 setup.py test -P visualization --remote-data=astropy --open-files -a "--mpl --mpl-results-path=$PWD/results"
+          command: |
+            pip install pytest-astropy-header --upgrade --no-deps
```
Comment (bsipocz) [inline .circleci/config.yml]: why is this necessary, isn't it added to `pytest-astropy`? Or just a temporary measure for now until there is a new upstream release?

File: CHANGES.rst [hunk 2]
```diff
@@ -179,6 +179,11 @@ astropy.table
 astropy.tests
 ^^^^^^^^^^^^^
 
+- The plugin that handles the custom header in the test output has been
+  moved to the pytest-astropy-header plugin package. See the README at
```
Comment (bsipocz) [inline CHANGES.rst]:
```suggestion
  moved to the ``pytest-astropy-header plugin`` package. See the README at
```

File: .circleci/config.yml (hunk 1 above)
Comment (astrofrog) [reply .circleci/config.yml]: It's temporary because pytest-astropy is already installed in the docker images but is an old version..

**Comment by astrofrog** [comment]
> We'll need to open batch PRs to packages that import display stuff from astropy, as this current deprecation is only a deprecation in name with no functionality behind, right?

Yes but just to be clear nothing will break badly, some packages might not appear anymore in the header, so it's not something we have to do urgently.

**Comment by bsipocz** [comment]
Oh, yes, certainly that's true. 
I would aim to open those PRs before release proper, but after RC, and it can easily be done as a hack at the coordination meeting.

**Comment by bsipocz** [comment]
I don't think we need to wait CI here, I go ahead and merge it now.

Thank you @astrofrog!

**Comment by pllim** [comment]
I don't think the warning can be fixed upstream. @jdavies-st said it's something specific to how it is defined in the core. Even though it is deprecated, it is still there. Maybe we closed #9437 too soon.

**Comment by bsipocz** [comment]
OK. Let's open  an issue for it upstream, that will be better contained and higher chance that we get it properly investigated.

**Comment by mhvk** [comment]
Hmm, now I cannot run tests any more:
```
ImportError while loading conftest '/tmp/user/1000/astropy-test-wjn4gd0d/lib/python3.7/site-packages/astropy/conftest.py'.
astropy/conftest.py:11: in <module>
    from pytest_astropy_header.display import PYTEST_HEADER_MODULES
E   ModuleNotFoundError: No module named 'pytest_astropy_header'
```
I tried a new clean `venv` and `git clean`, but am still using system packages (which really should work even if things were updated...)

**Comment by bsipocz** [comment]
We've managed to mess it up a bit, please install `pytest-astropy-header`, that will solve the issue.

**Comment by mhvk** [comment]
`pip install -U pytest-astropy` indeed solved it - good enough for now, thanks!
